### ISSUE TITLE
Add utils for pre-processing of data

### ISSUE THREAD

**Issue by abuzarmahmood** [abuzarmahmood/blechRNN#16]
- Binning input data
- Generating inputs and labels
- Setting up train and test sets

Refer to processing in: https://github.com/katzlabbrandeis/blech_clust/blob/master/utils/infer_rnn_rates.py

**Comment by abuzarmahmood** [comment]
### Summary of Suggestions and Changes

1. **Files Suggested for Modification:**
   - **`/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/model.py`**
     - **Description:** Related to the model setup for recurrent neural networks, specifically for firing rate estimation. It mentions inputs like spike trains with binning.
   - **`/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/run_model.py`**
     - **Description:** Responsible for running the model on data and generating outputs. It includes data loading and might be suitable for integrating data pre-processing steps.
   - **`/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/train.py`**
     - **Description:** Handles model training and prediction. Relevant for generating inputs and labels, and incorporating pre-processing steps.

2. **Specific Changes Proposed:**

   - **Binning Input Data:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/run_model.py`
     - **Line:** 42-47
     - **Description:** Implement binning of spikes by reshaping and summing over the last axis.
     - **Code Snippet:**
       ```python
       # Bin spikes
       bin_size = 25
       # (tastes x trials, neurons, time)
       # for example : (120, 35, 280)
       binned_spikes = np.reshape(cat_spikes, 
                                  (*cat_spikes.shape[:2], -1, bin_size)).sum(-1)
       ```

   - **Generating Inputs and Labels:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/run_model.py`
     - **Line:** 109-115
     - **Description:** Generate inputs and labels by shifting inputs to predict the next time-bin.
     - **Code Snippet:**
       ```python
       # Instead of predicting activity in the SAME time-bin,
       # predict activity in the NEXT time-bin
       # Forcing the model to learn temporal dependencies
       inputs_plus_context = inputs_plus_context[:-1]
       inputs = inputs[1:]
       
       # (seq_len * batch, output_size)
       labels = torch.from_numpy(inputs).type(torch.float32)
       # (seq_len, batch, input_size)
       inputs = torch.from_numpy(inputs_plus_context).type(torch.float)
       ```

   - **Setting Up Train and Test Sets:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/run_model.py`
     - **Line:** 117-134
     - **Description:** Implement a train-test split to separate data into training and testing datasets.
     - **Code Snippet:**
       ```python
       # Split into train and test
       train_test_split = 0.75
       train_inds = np.random.choice(
               np.arange(inputs.shape[1]), 
               int(train_test_split * inputs.shape[1]), 
               replace = False)
       test_inds = np.setdiff1d(np.arange(inputs.shape[1]), train_inds)
       
       train_inputs = inputs[:,train_inds]
       train_labels = labels[:,train_inds]
       test_inputs = inputs[:,test_inds]
       test_labels = labels[:,test_inds]
       
       train_inputs = train_inputs.to(device)
       train_labels = train_labels.to(device)
       test_inputs = test_inputs.to(device)
       test_labels = test_labels.to(device)
       ```

These changes aim to address the issue by adding necessary utilities for pre-processing data, including binning, generating inputs and labels, and setting up train and test sets.

---
*This response was automatically generated by blech_bot using model gpt-4o*

**Comment by abuzarmahmood** [comment]
### Summary of Suggestions and Changes

1. **Identifying Files for Modification:**
   - **`get_data.py`:** Responsible for data extraction and potentially binning spike-trains.
   - **`train.py`:** Involved in model training and prediction, including setting up train and test sets.
   - **`model.py`:** Deals with the recurrent neural network, requiring pre-processed inputs.
   - **`run_model.py`:** Used to run the model on data, involving pre-processed data for execution.

2. **Specific Changes Suggested:**

   - **Binning Input Data:**
     - **File:** `get_data.py`
     - **Lines:** 2-3
     - **Description:** Implement a function to bin spike-trains.
     - **Code Snippet:**
       ```python
       def bin_spike_trains(spike_trains, bin_width):
           # Reshape and sum over the bin width
           binned_spikes = np.reshape(spike_trains, (*spike_trains.shape[:2], -1, bin_width)).sum(-1)
           return binned_spikes
       ```

   - **Generating Inputs and Labels:**
     - **File:** `run_model.py`
     - **Lines:** 42-115
     - **Description:** Use binned spike data to generate inputs and labels.
     - **Code Snippet:**
       ```python
       # Bin spikes
       bin_size = 25
       binned_spikes = bin_spike_trains(cat_spikes, bin_size)

       # Reshape to (seq_len, batch, input_size)
       inputs = binned_spikes.copy()
       inputs = np.moveaxis(inputs, -1, 0)

       # Generate labels by shifting inputs
       labels = inputs[1:]
       inputs = inputs[:-1]
       ```

   - **Setting Up Train and Test Sets:**
     - **File:** `run_model.py`
     - **Lines:** 117-134
     - **Description:** Split data into training and testing sets.
     - **Code Snippet:**
       ```python
       # Split into train and test
       train_test_split = 0.75
       train_inds = np.random.choice(
           np.arange(inputs.shape[1]), 
           int(train_test_split * inputs.shape[1]), 
           replace=False
       )
       test_inds = np.setdiff1d(np.arange(inputs.shape[1]), train_inds)

       train_inputs = inputs[:, train_inds]
       train_labels = labels[:, train_inds]
       test_inputs = inputs[:, test_inds]
       test_labels = labels[:, test_inds]
       ```

These changes aim to address the issue by implementing utilities for pre-processing data, including binning input data, generating inputs and labels, and setting up train and test sets.

---
*This response was automatically generated by blech_bot using model gpt-4o*
